Mathlib/GroupTheory/RegularWreathProduct.lean

```diff
@@ -0,0 +1,301 @@
+  one_smul := by simp
+  mul_smul := by simp [smul_smul, mul_assoc]
+
+variable [FaithfulSMul D Λ]
+instance instFaithfulSMulRWP [Nonempty Q] [Nonempty Λ] : FaithfulSMul (D ≀ᵣ Q) (Λ × Q) where
+  eq_of_smul_eq_smul := by
+    simp only [rsmul, Prod.mk.injEq, mul_left_inj, Prod.forall]
+    intro m₁ m₂ h
+    let ⟨a⟩ := ‹Nonempty Λ›
+    let ⟨b⟩ := ‹Nonempty Q›
+    ext q
+    · have hh := fun a => (h a (m₁.right⁻¹ * q)).1
+      rw [← (h a b).2] at hh
+      group at hh
+      exact FaithfulSMul.eq_of_smul_eq_smul hh
+    · exact (h a b).2
+
+/-- The map sending the wreath product `D ≀ᵣ Q` to its representation as a permutation of `Λ × Q`
+given `D`-set `Λ`. -/
+def toPerm : D ≀ᵣ Q →* Equiv.Perm (Λ × Q) :=
+  MulAction.toPermHom (D ≀ᵣ Q) (Λ × Q)
+
+theorem toPermInj [Nonempty Λ] : Function.Injective (toPerm D Q Λ) := MulAction.toPerm_injective
+
+end perm
+
+end RegularWreathProduct
+section iterated
+
+universe u
+
+/-- The wreath product of group `G` iterated `n` times. -/
+def IteratedWreathProduct (G : Type u) : (n : ℕ) → Type u
+| 0 => PUnit
+| n + 1 => (IteratedWreathProduct G n) ≀ᵣ G
+
+variable (G : Type u) (n : ℕ)
+
+@[simp]
+lemma IteratedWreathProduct_zero : IteratedWreathProduct G 0 = PUnit := rfl
+
+@[simp]
+lemma IteratedWreathProduct_succ :
+    IteratedWreathProduct G (n + 1) = (IteratedWreathProduct G n) ≀ᵣ G := rfl
+
+instance [Finite G] : Finite (IteratedWreathProduct G n) := by
+  induction n with
+  | zero => rw [IteratedWreathProduct_zero]; infer_instance
+  | succ n h => rw [IteratedWreathProduct_succ]; infer_instance
+
+theorem iter_wreath_card [Finite G] : Nat.card (IteratedWreathProduct G n) =
+    Nat.card G ^ (∑ i ∈ Finset.range n, Nat.card G ^ i) := by
+  induction n with
+  | zero => simp
+  | succ n h => rw [IteratedWreathProduct_succ, RegularWreathProduct.card,
+      h, geom_sum_succ, pow_succ, pow_mul']
+
+variable [Group G]
+
+instance : Group (IteratedWreathProduct G n) := by
+ induction n with
+ | zero => rw [IteratedWreathProduct_zero]; infer_instance
+ | succ n ih => rw [IteratedWreathProduct_succ]; infer_instance
+
+lemma mu_eq {p n : ℕ} [hp : Fact (Nat.Prime p)] :
+    (p ^ n).factorial.factorization p = ∑ i ∈ Finset.range n, p ^ i := by
+  rw [← Nat.multiplicity_eq_factorization hp.out (p ^ n).factorial_ne_zero, ← ENat.coe_inj,
+    ← (Nat.finiteMultiplicity_iff.2
+      ⟨hp.out.ne_one, (p ^ n).factorial_pos⟩).emultiplicity_eq_multiplicity]
+  induction n with
+  | zero => simp [hp.out.emultiplicity_one]
+  | succ n h =>
+    rw [pow_succ', hp.out.emultiplicity_factorial_mul, h, Finset.sum_range_succ, ENat.coe_add]
+
+/-- If `α` is equivalent to `β`, then `Perm α` is isomorphic to `Perm β`. -/
+def Equiv.permCongrHom {α β : Type*} (e : α ≃ β) : Equiv.Perm α ≃* Equiv.Perm β where
+  toFun x := e.symm.trans (x.trans e)
+  invFun y := e.trans (y.trans e.symm)
+  left_inv _ := by ext; simp
+  right_inv _ := by ext; simp
+  map_mul' _ _ := by ext; simp
```

Comment:
Can find a better home for this definition? (Maybe `Mathlib.Algebra.Group.End` where the group structure on `Equiv.Perm` is defined?)